install/tools/istio_vm_common.sh

```diff
@@ -0,0 +1,216 @@
+  echo "address=/istio-pilot/$PILOT_IP" >> kubedns
+  echo "address=/istio-ca/$CA_IP" >> kubedns
+
+  CIDR=$(gcloud container clusters describe ${K8SCLUSTER} $(_istioGcloudOpt) --format "value(servicesIpv4Cidr)")
+  echo "ISTIO_SERVICE_CIDR=$CIDR" > cluster.env
+}
+
+
+# Install required files on a VM and run the setup script.
+#
+# Must be run for each VM added to the cluster
+# Params:
+# - name of the VM - used to copy files over.
+# - service account to be provisioned (defaults to istio.default)
+function istioProvisionVM() {
+ NAME=${1}
+
+ local SA=${2:-istio.default}
+
+  kubectl get secret $SA -o jsonpath='{.data.cert-chain\.pem}' |base64 -d  > cert-chain.pem
+  kubectl get secret $SA -o jsonpath='{.data.root-cert\.pem}' |base64 -d  > root-cert.pem
+  kubectl get secret $SA -o jsonpath='{.data.key\.pem}' |base64 -d  > key.pem
+
+  local ISTIO_FILES=${ISTIO_FILES:-.}
+
+ # Copy deb, helper and config files
+ # Reviews not copied - VMs don't support labels yet.
+ istioCopy $NAME \
+   kubedns \
+   *.pem \
+   cluster.env \
+   $ISTIO_FILES/istio_vm_setup.sh \
+   $ISTIO_FILES/istio-proxy-envoy.deb \
+   $ISTIO_FILES/istio-agent.deb \
+   $ISTIO_FILES/istio-auth-node-agent.deb \
+
+ # Run the setup script.
+ istioRun $NAME "sudo bash -c -x ./istio_vm_setup.sh"
+}
+
+
+# Helper functions for the main script
+
+# If Istio was built from source, copy the artifcats to the current directory, for use
+# by istioProvisionVM
+function istioCopyBuildFiles() {
+  local ISTIO_IO=${ISTIO_BASE:-${GOPATH:-$HOME/go}}/src/istio.io
+  local ISTIO_FILES=${ISTIO_FILES:-.}
+
+  (cd $ISTIO_IO/proxy; bazel build tools/deb/... )
+  (cd $ISTIO_IO/pilot; bazel build tools/deb/... )
+  (cd $ISTIO_IO/auth; bazel build tools/deb/... )
+
+  cp $ISTIO_IO/proxy/bazel-bin/tools/deb/istio-proxy-envoy.deb \
+     $ISTIO_IO/pilot/bazel-bin/tools/deb/istio-agent.deb \
+     $ISTIO_IO/auth/bazel-bin/tools/deb/istio-auth-node-agent.deb \
+     $ISTIO_IO/istio/install/tools/istio_vm_setup.sh \
+     $ISTIO_FILES
+  # For override to work
+  chmod +w *.deb
+}
+
+# Copy files to the VM
+function istioCopy() {
+  # TODO: based on some env variable, use different commands for other clusters or for testing with
+  # bare-metal machines.
+  local NAME=$1
+  shift
+  local FILES=$*
+
+  gcloud compute scp --recurse $(_istioGcloudOpt) $FILES ${NAME}:
+}
+
+# Run a command in a VM.
+function istioRun() {
+  local NAME=$1
+  local CMD=$2
+
+  gcloud compute ssh $(_istioGcloudOpt) $NAME --command "$CMD"
+}
+
+# Helper to generate options for gcloud compute.
+# PROJECT and ISTIO_ZONE are optional. If not set, the defaults will be used.
+#
+# Example default setting:
+# gcloud config set compute/zone us-central1-a
+# gcloud config set core/project costin-raw
+function _istioGcloudOpt() {
+    local OPTS=""
+    if [[ ${PROJECT:-} != "" ]]; then
+       OPTS="--project $PROJECT"
+    fi
+    if [[ ${ISTIO_ZONE:-} != "" ]]; then
+       OPTS="$OPTS --zone $ISTIO_ZONE"
+    fi
+    echo $OPTS
```

Comment:
Is it possible to have cloud platform dependent methods in one file vs spread out?  That seems easier to maintain.

Reply:
I can try to group them, technically I can get rid of this function and rely on defaults. 
I suppose I can also use env variables to inject 'gcloud' if it is a major concern - we can 
refine it as we test on more platforms.

Reply:
Maybe in a separate PR ?